packages/marshal/src/encodeToSmallcaps.js

```diff
@@ -0,0 +1,461 @@
+ * @property {(promise: object) => SmallcapsEncoding} [encodePromiseToSmallcaps]
+ * @property {(error: object) => SmallcapsEncoding} [encodeErrorToSmallcaps]
+ */
+
+const dontEncodeRemotableToSmallcaps = rem =>
+  assert.fail(X`remotable unexpected: ${rem}`);
+
+const dontEncodePromiseToSmallcaps = prom =>
+  assert.fail(X`promise unexpected: ${prom}`);
+
+const dontEncodeErrorToSmallcaps = err =>
+  assert.fail(X`error object unexpected: ${q(err)}`);
+
+/**
+ * @param {EncodeToSmallcapsOptions=} encodeOptions
+ * @returns {(passable: Passable) => SmallcapsEncoding}
+ */
+export const makeEncodeToSmallcaps = ({
+  encodeRemotableToSmallcaps = dontEncodeRemotableToSmallcaps,
+  encodePromiseToSmallcaps = dontEncodePromiseToSmallcaps,
+  encodeErrorToSmallcaps = dontEncodeErrorToSmallcaps,
+} = {}) => {
+  /**
+   * Must encode `val` into plain JSON data *canonically*, such that
+   * `JSON.stringify(encode(v1)) === JSON.stringify(encode(v1))`. For most
+   * encodings, the order of properties of each node of the output
+   * structure is determined by the algorithm below without special
+   * arrangement, usually by being expressed directly as an object literal.
+   * The exception is copyRecords, whose natural enumeration order
+   * can differ between copyRecords that our distributed object semantics
+   * considers to be equivalent.
+   * Since, for each copyRecord, we only accept string property names,
+   * not symbols, we can canonically sort the names first.
+   * JSON.stringify will then visit these in that sorted order.
+   *
+   * SmallcapsEncoding with a canonical-JSON encoder would also solve
+   * this canonicalness
+   * problem in a more modular and encapsulated manner. Note that the
+   * actual order produced here, though it agrees with canonical-JSON on
+   * copyRecord property ordering, differs from canonical-JSON as a whole
+   * in that the other record properties are visited in the order in which
+   * they are literally written below. TODO perhaps we should indeed switch
+   * to a canonical JSON encoder, and not delicatetly depend on the order
+   * in which these object literals are written.
+   *
+   * Readers must not care about this order anyway. We impose this requirement
+   * mainly to reduce non-determinism exposed outside a vat.
+   *
+   * @param {Passable} passable
+   * @returns {SmallcapsEncoding} except that `encodeToSmallcaps` does not generally
+   * `harden` this result before returning. Rather, `encodeToSmallcaps` is not
+   * directly exposed.
+   * What's exposed instead is a wrapper that freezes the output before
+   * returning. If this turns out to impede static analysis for `harden` safety,
+   * we can always put the (now redundant) hardens back in. They don't hurt.
+   */
+  const encodeToSmallcapsRecur = passable => {
+    // First we handle all primitives. Some can be represented directly as
+    // JSON, and some must be encoded as [SCLASS] composites.
```

Comment:
This comment is no longer accurate as SCLASS is not not the only encoding for non-JSON serializable data.

Reply:
Good catch. How about
```js
    // First we handle all primitives. Some can be represented directly as
    // JSON, and some must be encoded into smallcaps strings.
```
?
For all primitives, smallcaps uses only strings.

Reply:
Resolving.